### PR TITLE
Support custom encoding handler

### PR DESCRIPTION
**Pull request by ITJesse**
The default encoding is utf-8 which always failed on most of Japanese website which using JIS.

### CODE REVIEW

**Comment by laurengarcia** [comment]
Looks good, will pull this down and test a bit.

**Comment by laurengarcia** [comment]
Will you send me an example of how you would use this? I can't figure out what your `options.decoder` would look like; it seems like its not any of the options available from the Node.js Buffer api. 

Please forgive my ignorance of Japanese encoding.

**Comment by laurengarcia** [comment]
Leaving this here for reference, it is a similar discussion from the `request` module:
https://github.com/request/request/issues/1080#issuecomment-56172161

File: index.js [hunk 1]
```diff
@@ -37,6 +38,9 @@ module.exports = function (url, options) {
       // rewrite url if our request had to follow redirects to resolve the
       // final link destination (for example: links shortened by bit.ly)
       if (response.request.uri.href) url = response.request.uri.href
+      if (opts.decoder) {
+        body = opts.decoder(body)
+      }
```
Comment (laurengarcia) [inline index.js]: If I'm reading this correctly, the decoder would need to accept the buffer returned from the `request` module and convert it with `opts.decoder` before parsing the metadata. Something like one of the following packages might work(?): 

https://www.npmjs.com/package/encoding-japanese
https://www.npmjs.com/package/iconv-lite

File: index.js [hunk 2]
```diff
@@ -13,20 +13,21 @@ module.exports = function (url, options) {
...
     url: url,
     headers: {
       'User-Agent': opts.userAgent,
-      'From': opts.fromEmail
+      From: opts.fromEmail,
     },
     maxRedirects: opts.maxRedirects,
-    encoding: 'utf8',
-    timeout: opts.timeout
+    encoding: opts.decoder ? null : 'utf8',
```
[the remark this answers is not in the document]
Comment (ITJesse) [reply index.js]: encoding - encoding to be used on setEncoding of response data. If null, the body is returned as a Buffer. Anything else (including the default value of undefined) will be passed as the encoding parameter to toString() (meaning this is effectively utf8 by default). (Note: if you expect binary data, you should set encoding: null.)
[https://github.com/request/request](url)

File: index.js (hunk 1 above)
Comment (ITJesse) [reply index.js]:
```typescript
import detectCharacterEncoding from 'detect-character-encoding'
import iconv from 'iconv-lite'

const decoder = val => {
  const result = detectCharacterEncoding(val)
  if (!result) {
    return ''
  }
  const { encoding } = result
  if (encoding === 'UTF-8') {
    return val.toString()
  }
  if (!iconv.encodingExists(encoding)) {
    return ''
  }
  return iconv.decode(val, encoding)
}
```

Comment (laurengarcia) [reply index.js]: Ok perfect, that was what i needed. Will try this tomorrow.

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (laurengarcia) [reply index.js]: My bad, i misread this. Sorry!

**Comment by laurengarcia** [comment]
Ended up merging this in pieces. 
v2.3: options.decode() to handle custom encodings
v2.4: Typescript definitions index.d.ts

https://github.com/LevelNewsOrg/url-metadata/blob/master/CHANGELOG.md

Published to npm:
https://www.npmjs.com/package/url-metadata

Thank-you!

**Pull request closed without merging**